File: cotsSource/org.apache.qpid/java/management/console/src/main/java/org/apache/qpid/console/QMFObject.java
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */
package org.apache.qpid.console;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;
import java.util.Map.Entry;

import org.apache.qpid.transport.codec.Decoder;
import org.apache.qpid.transport.codec.Encoder;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import edu.emory.mathcs.backport.java.util.Arrays;

public class QMFObject
{
    private static Logger log = LoggerFactory.getLogger(QMFObject.class);
    protected SchemaClass schema;
    private java.util.Date createTime;
    private java.util.Date currentTime;
    private java.util.Date deleteTime;
    private ObjectID objectID;
    private Session session;
    private boolean managed;
    public java.util.HashMap<String, Object> properties = new java.util.HashMap<String, Object>();
    public java.util.HashMap<String, Object> statistics = new java.util.HashMap<String, Object>();

    // This constructor is the "naked" constructor which creates
    // an object without a session or a schema. It is used by
    // subclasses which are auto generated
    public QMFObject()
    {
    }

    public QMFObject(QMFObject source)
    {
        this.setSession(source.getSession());
        this.setSchema(source.getSchema());
        this.managed = source.managed;
        this.setCurrentTime(source.getCurrentTime());
        this.setCreateTime(source.getCreateTime());
        this.setDeleteTime(source.getDeleteTime());
        this.setObjectID(source.getObjectID());
        this.properties = source.properties;
        this.statistics = source.statistics;
    }

    // This constructor is used by a session make object call to
    // create a blank object from a schema.
    public QMFObject(Session session, SchemaClass schema,
            boolean hasProperties, boolean hasStats, boolean isManaged)
    {
        setSession(session);
        setSchema(schema);
        managed = isManaged;
        if (hasProperties)
        {
            for (SchemaProperty prop : getSchema().getAllProperties())
            {
                Object propValue = null;
                if (!prop.getOptional())
                {
                    propValue = Util.defaultValue(prop.getType());
                }
                this.setProperty(prop.getName(), propValue);
            }
        }
        if (hasStats)
        {
            for (SchemaStatistic stat : getSchema().statistics)
            {
                setStatistic(stat.getName(), Util.defaultValue(stat.getType()));
            }
        }
    }

    // This constructor is used by the session to create an object based on a
    // data
    // stream by the agent.
    public QMFObject(Session session, SchemaClass schema, Decoder dec,
            boolean hasProperties, boolean hasStats, boolean isManaged)
    {
        setSession(session);
        setSchema(schema);
        managed = isManaged;
        if (managed)
        {
            // FIXME DateTime or Uint64??
            setCurrentTime(new java.util.Date(dec.readDatetime()));
            setCreateTime(new java.util.Date(dec.readDatetime()));
            setDeleteTime(new java.util.Date(dec.readDatetime()));
            setObjectID(new ObjectID(dec));
        }
        if (hasProperties)
        {
            java.util.ArrayList<String> excluded = processPresenceMasks(dec,
                    getSchema());
            for (SchemaProperty prop : getSchema().getAllProperties())
            {
                if (excluded.contains(prop.getName()))
                {
                    // log.Debug(String.Format("Setting Property Default {0}",
                    // prop.Name)) ;
                    safeAddProperty(prop.getName(), null);
                } else
                {
                    // log.Debug(String.Format("Setting Property {0}",
                    // prop.Name)) ;
                    safeAddProperty(prop.getName(), session.decodeValue(dec,
                            prop.getType()));
                }
            }
        }
        if (hasStats)
        {
            for (SchemaStatistic stat : getSchema().getAllStatistics())
            {
                // log.Debug(String.Format("Setting Statistic {0}", stat.Name))
                // ;
                statistics.put(stat.getName(), session.decodeValue(dec, stat
                        .getType()));
            }
        }
    }

    public final long agentBank()
    {
        return getObjectID().agentBank();
    }

    public final long brokerBank()
    {
        return getObjectID().brokerBank();
    }

    public final void encode(Encoder enc)
    {
        int mask = 0;
        int bit = 0;
        java.util.ArrayList<SchemaProperty> propsToEncode = new java.util.ArrayList<SchemaProperty>();
        log.debug(String.format("Encoding class %s:%s", getSchema()
                .getPackageName(), getSchema().getClassName()));
        enc.writeUint8((short) 20);
        getSchema().getKey().encode(enc);
        for (SchemaProperty prop : getSchema().getAllProperties())
        {
            if (prop.getOptional())
            {
                if (bit == 0)
                {
                    bit = 1;
                }
                if ((properties.containsKey(prop.getName()))
                        && (properties.get(prop.getName()) != null))
                {
                    mask |= bit;
                    propsToEncode.add(prop);
                } else
                {
                }
                bit = bit << 1;
                if (bit == 256)
                {
                    bit = 0;
                    enc.writeUint8((short) mask);
                    mask = 0;
                }
            } else
            {
                propsToEncode.add(prop);
            }
        }
        if (bit != 0)
        {
            enc.writeUint8((short) mask);
        }
        for (SchemaProperty prop : propsToEncode)
        {
            Object obj = properties.get(prop.getName());
            // log.Debug(String.Format("Encoding property {0}", prop.Name)) ;
            getSession().encodeValue(enc, prop.getType(), obj);
        }
        for (SchemaStatistic stat : getSchema().statistics)
        {
            Object obj = statistics.get(stat.getName());
            getSession().encodeValue(enc, stat.getType(), obj);
        }
        log.debug("Done");
    }

    public final Date getCreateTime()
    {
        return createTime;
    }

    public final Date getCurrentTime()
    {
        return currentTime;
    }

    public final Date getDeleteTime()
    {
        return deleteTime;
    }

    protected final ArrayList<SchemaMethod> getMethods()
    {
        return getSchema().getAllMethods();
    }

    public final ObjectID getObjectID()
    {
        return objectID;
    }

    public final Object getProperty(String attributeName)
    {
        return properties.get(attributeName);
    }

    public SchemaClass getSchema()
    {
        return schema;
    }

    public final Session getSession()
    {
        return session;
    }

    protected final MethodResult internalInvokeMethod(String name,
            List<Object> args, boolean synchronous, int timeToLive)
    {
        if (!managed)
        {
            throw new ConsoleException("Object is not Managed");
        }
        if (getSchema().getMethod(name) == null)
        {
            throw new ConsoleException(String.format(
                    "Method named '%s' does not exist", name));
        }
        return getSession().invokeMethod(this, name, args, synchronous,
                timeToLive);
    }

    public final MethodResult invokeMethod(String name, boolean synchronous,
            int timeToLive, Object... args)
    {
        return this.internalInvokeMethod(name, Arrays.asList(args),
                synchronous, timeToLive);
    }

    public final MethodResult invokeMethod(String name, boolean synchronous,
            Object... args)
    {
        return this.internalInvokeMethod(name, Arrays.asList(args),
                synchronous, Broker.SYNC_TIME);
    }

    public final MethodResult invokeMethod(String name, int timeToLive,
            Object... args)
    {
        return this.internalInvokeMethod(name, Arrays.asList(args), true,
                timeToLive);
    }

    public final MethodResult invokeMethod(String name, Object... args)
    {
        return this.internalInvokeMethod(name, Arrays.asList(args), true,
                Broker.SYNC_TIME);
    }

    public final boolean isDeleted()
    {
        return !getDeleteTime().equals(new java.util.Date(0));
    }

    protected final ArrayList<String> processPresenceMasks(Decoder dec,
            SchemaClass schema)
    {
        java.util.ArrayList<String> excludes = new java.util.ArrayList<String>();
        short bit = 0;
        short mask = 0;
        for (SchemaProperty prop : getSchema().getAllProperties())
        {
            if (prop.getOptional())
            {
                // log.Debug(String.Format("Property named {0} is optional",
                // prop.Name)) ;
                if (bit == 0)
                {
                    mask = dec.readUint8();
                    bit = 1;
                }
                if ((mask & bit) == 0)
                {
                    // log.Debug(String.Format("Property named {0} is not present",
                    // prop.Name)) ;
                    excludes.add(prop.getName());
                }
                bit *= 2;
                if (bit == 256)
                {
                    bit = 0;
                }
            }
        }
        return excludes;
    }

    public final String routingKey()
    {
        return getObjectID().routingCode();
    }

    protected final void safeAddProperty(String propName, Object value)
    {
        if (properties.containsKey(propName))
        {
            properties.put(propName, value);
        } else
        {
            properties.put(propName, value);
        }
    }

    public final void setCreateTime(java.util.Date value)
    {
        createTime = value;
    }

    public final void setCurrentTime(java.util.Date value)
    {
        currentTime = value;
    }

    public final void setDeleteTime(java.util.Date value)
    {
        deleteTime = value;
    }

    public final void setObjectID(ObjectID value)
    {
        objectID = value;
    }

    public final void setProperty(String attributeName, Object newValue)
    {
        properties.put(attributeName, newValue);
    }

    public void setSchema(SchemaClass value)
    {
        schema = value;
    }

    public final void setSession(Session value)
    {
        session = value;
    }

    protected final void setStatistic(String attributeName, Object newValue)
    {
        statistics.put(attributeName, newValue);
    }

    @Override
    public String toString()
    {
        String propertyString = "";
        for (Entry<String, Object> pair : properties.entrySet())
        {
            propertyString = propertyString
                    + String.format("(Name: '%0$s' Value: '%1$s')", pair
                            .getKey(), pair.getValue());
        }
        String statsString = "";
        for (Entry<String, Object> sPair : statistics.entrySet())
        {
            statsString = statsString
                    + String.format("(Name: '%0$s' Value: '%1$s')", sPair
                            .getKey(), sPair.getValue());
        }
        if (managed)
        {
            return String
                    .format(
                            "Managed QMFObject %0$s:%1$s(%2$s) Properties: [%3$s] Statistics: [%4$s])",
                            getSchema().getPackageName(), getSchema()
                                    .getClassName(), getObjectID(),
                            propertyString, statsString);
        } else
        {
            return String
                    .format(
                            "QMFObject %0$s:%1$s Properties: [%2$s] Statistics: [%3$s]",
                            getSchema().getPackageName(), getSchema()
                                    .getClassName(), propertyString,
                            statsString);
        }
    }
}

File: cotsSource/org.apache.qpid/dotnet/client-010/management/console/QMFObject.cs
/*
 *
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0
 * 
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 */

using System;
using System.Collections.Generic ;
using log4net ;
using org.apache.qpid.transport.codec ;

namespace org.apache.qpid.console
{
	
	/**
	 * An object which is returned from an agent by the Session. It can have
	 * methods, properties, and statistics.
	 */
	public class QMFObject
	{
	
		public static ILog log = LogManager.GetLogger(typeof(QMFObject)) ;	

 		public Session Session  {get;set;}
 		protected SchemaClass _Schema ;
 		virtual public SchemaClass Schema {get {return _Schema;} set {_Schema=value;}}
 		bool Managed ;
 		public DateTime CurrentTime {get;set;}
		public DateTime CreateTime {get;set;} 		
		public DateTime DeleteTime {get;set;}		
		public ObjectID ObjectID {get;set;}
		public Dictionary<string, object> Properties = new Dictionary<string, object>() ;
		public Dictionary<string, object> Statistics = new Dictionary<string, object>() ;		
			
 		
 		// This constructor is the "naked" constructor which creates
 		// an object without a session or a schema. It is used by
 		// subclasses which are auto generated
 		public QMFObject() {
 		}
 		
 		public QMFObject(QMFObject source) {
 			this.Session = source.Session ;
 			this.Schema = source.Schema ;
 			this.Managed = source.Managed ;
 			this.CurrentTime = source.CurrentTime ;
 			this.CreateTime = source.CreateTime ;
 			this.DeleteTime = source.DeleteTime ;
 			this.ObjectID = source.ObjectID ;
 			this.Properties = source.Properties ;
 			this.Statistics = source.Statistics ;
 		}
 		
 		// This constructor is used by a session make object call to 
 		// create a blank object from a schema.
 		public QMFObject(Session session, SchemaClass schema, bool hasProperties, bool hasStats , bool isManaged) {
			Session = session ;
			Schema = schema ;
			Managed = isManaged ; 
			
			if (hasProperties) {
				foreach (SchemaProperty prop in Schema.GetAllProperties()) {
					object propValue = null ; 
					if (!prop.Optional) {
						propValue = Util.DefaultValue(prop.Type) ;
					}
					this.SetProperty(prop.Name, propValue) ;			
				}
			}	
			
			if (hasStats) {
				foreach (SchemaStatistic stat in Schema.Statistics)  {			
					SetStatistic(stat.Name, Util.DefaultValue(stat.Type)) ;
				}
			}															
 		}
 		
 		// This constructor is used by the session to create an object based on a data 
 		// stream by the agent.
		public QMFObject(Session session, SchemaClass schema, IDecoder dec, bool hasProperties, bool hasStats , bool isManaged)
		{
			Session = session ;
			Schema = schema ;
			Managed = isManaged ;
			
			if (Managed) {
			    // FIXME DateTime or Uint64??
				CurrentTime = new DateTime(dec.ReadDatetime()) ;
				CreateTime = new DateTime(dec.ReadDatetime()) ;				
				DeleteTime = new DateTime(dec.ReadDatetime()) ;				
				ObjectID = new ObjectID(dec) ;
			}
			
			if (hasProperties) {
				List<string> excluded = ProcessPresenceMasks(dec, Schema) ;
				
				foreach (SchemaProperty prop in Schema.GetAllProperties()) {
					if (excluded.Contains(prop.Name)) {
					    log.Debug(String.Format("Setting Property Default {0}", prop.Name)) ;					    					
						safeAddProperty(prop.Name, null) ;	
					} else {
						//log.Debug(String.Format("Setting Property {0}", prop.Name)) ;
						safeAddProperty(prop.Name, session.DecodeValue(dec, prop.Type)) ;
					}
				}
			}
			
			if (hasStats) {
				foreach (SchemaStatistic stat in Schema.GetAllStatistics())  {
					//log.Debug(String.Format("Setting Statistic {0}", stat.Name)) ;				
					Statistics.Add(stat.Name, session.DecodeValue(dec, stat.Type)) ;
				}
			}
			
		}

		protected List<string> ProcessPresenceMasks(IDecoder dec, SchemaClass schema) {
			List<string> excludes = new List<string> () ;
			short bit = 0 ;
			short mask = 0 ;
			foreach (SchemaProperty prop in Schema.GetAllProperties()) {
				if (prop.Optional) {
					//log.Debug(String.Format("Property named {0} is optional", prop.Name)) ;				
					if (bit == 0) {
						mask=dec.ReadUint8() ;
						bit = 1 ;
					}

					if ((mask & bit) == 0) {
						//log.Debug(String.Format("Property named {0} is not present", prop.Name)) ;
						excludes.Add(prop.Name) ;
					}
					bit *= 2 ;
					if (bit == 256) {
						bit = 0 ;
					}
				}
			}
			return excludes ;			
		}
		
		protected List<SchemaMethod> getMethods() {
			return Schema.GetAllMethods() ;
		}
		
		public object GetProperty(string attributeName) {
			//FIXME any object refs?
			object returnValue = null ;
			Properties.TryGetValue(attributeName, out returnValue) ;
			return returnValue ;
		}
		
		protected void SetStatistic(string attributeName, object newValue) {
			Statistics.Remove(attributeName) ;
			Statistics.Add(attributeName, newValue) ;
		}
		
		public void SetProperty(string attributeName, object newValue) {
			Properties.Remove(attributeName) ;
			Properties.Add(attributeName, newValue) ;
		}		
		
		public bool IsDeleted() {
			return !DeleteTime.Equals(new DateTime(0)) ;
		}
		
		public string RoutingKey() {
			return ObjectID.RoutingCode() ;
		}
		
		public long AgentBank() {
			return ObjectID.AgentBank() ;
		}
		
		public long BrokerBank() {
			return ObjectID.BrokerBank() ;
		}
				
		override public string ToString() {
			string propertyString = "" ;
			foreach (KeyValuePair<string, object> pair in Properties) {
				propertyString = propertyString + String.Format("(Name: '{0}' Value: '{1}')", pair.Key, pair.Value) ;
			}
			
			string statsString = "" ;
			foreach (KeyValuePair<string, object> sPair in Statistics) {
				statsString = statsString + String.Format("(Name: '{0}' Value: '{1}')", sPair.Key, sPair.Value) ;
			}			
			if (Managed) {
				return String.Format("Managed QMFObject {0}:{1}({2}) Properties: [{3}] Statistics: [{4}])", Schema.PackageName, Schema.ClassName, ObjectID, propertyString, statsString) ;
			} else {
				return String.Format("QMFObject {0}:{1} Properties: [{2}] Statistics: [{3}]", Schema.PackageName, Schema.ClassName, propertyString, statsString) ;
			}
		}
		
		public MethodResult InvokeMethod(string name, params object[] args) {
			return this.InternalInvokeMethod(name, new List<object>(args), true, Broker.SYNC_TIME ) ;
		}
		
		public MethodResult InvokeMethod(string name, int timeToLive, params object[] args) {
			return this.InternalInvokeMethod(name, new List<object>(args), true, timeToLive) ;
		}
		
		public MethodResult InvokeMethod(string name, bool synchronous, int timeToLive, params object[] args) {
			return this.InternalInvokeMethod(name, new List<object>(args), synchronous, timeToLive ) ;
		}
				
		public MethodResult InvokeMethod(string name, bool synchronous, params object[] args) {
			return this.InternalInvokeMethod(name, new List<object>(args), synchronous, Broker.SYNC_TIME) ;
		}
		
		protected MethodResult InternalInvokeMethod(string name, List<object> args, bool synchronous, int timeToLive) {
			if (!Managed) {
				throw new Exception ("Object is not Managed") ;
			}
			if (Schema.GetMethod(name) == null) {
				throw new Exception (String.Format("Method named '{0}' does not exist", name))  ;
			}
			return Session.InvokeMethod(this, name, args, synchronous, timeToLive) ;
		}
		
		public void Encode (IEncoder enc) {
			int mask = 0 ;
			int bit = 0 ;
			List<SchemaProperty> propsToEncode = new List<SchemaProperty>() ;
			log.Debug(String.Format("Encoding class {0}:{1}", Schema.PackageName, Schema.ClassName)) ;
					
			enc.WriteUint8(20) ;
			Schema.Key.encode(enc) ;
			
			foreach (SchemaProperty prop in Schema.GetAllProperties()) {
				if (prop.Optional) {
					if (bit == 0) {
						bit =1 ;
					}
					if ((Properties.ContainsKey(prop.Name)) && (Properties[prop.Name] != null)) {
						mask |= bit ;
						propsToEncode.Add(prop) ;
					} else {
					}
					bit = bit << 1 ;
					if (bit == 256) {
						bit = 0 ;
						enc.WriteUint8((short)mask) ;
						mask = 0 ;
					}
				} else {
					propsToEncode.Add(prop) ;
				}
			}
			if (bit != 0) {
				enc.WriteUint8((short)mask) ;
			}		
			
			foreach (SchemaProperty prop in propsToEncode) {
				object obj = Properties[prop.Name] ;
				//log.Debug(String.Format("Encoding property {0}", prop.Name)) ;
				Session.EncodeValue(enc, prop.Type, obj) ;
			}
			foreach (SchemaStatistic stat in Schema.Statistics) {
				object obj = Statistics[stat.Name] ;
				Session.EncodeValue(enc, stat.Type, obj) ;
			}			
			log.Debug("Done") ;
		 
		 }
		 
		
		protected void safeAddProperty(string propName, object value) {
			if (Properties.ContainsKey(propName)) {
				Properties[propName] = value ;
			} else {
				Properties.Add(propName, value) ;
			}
		}
				 
		
	}
}
